eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs

```diff
@@ -144,3 +186,220 @@ fromGovActionStateSeq = Proposals . OMap.fromFoldable
+  toJSON = object . toPrevGovActionIdsChildrenPairs
+  toEncoding = pairs . mconcat . toPrevGovActionIdsChildrenPairs
+
+data PrevGovActionIds era = PrevGovActionIds
+  { pgaPParamUpdate :: !(StrictMaybe (PrevGovActionId 'PParamUpdatePurpose (EraCrypto era)))
+  -- ^ The last enacted GovActionId for a protocol parameter update
+  , pgaHardFork :: !(StrictMaybe (PrevGovActionId 'HardForkPurpose (EraCrypto era)))
+  -- ^ The last enacted GovActionId for a hard fork
+  , pgaCommittee :: !(StrictMaybe (PrevGovActionId 'CommitteePurpose (EraCrypto era)))
+  -- ^ The last enacted GovActionId for a committee change or no confidence vote
+  , pgaConstitution :: !(StrictMaybe (PrevGovActionId 'ConstitutionPurpose (EraCrypto era)))
+  -- ^ The last enacted GovActionId for a new constitution
+  }
+  deriving (Eq, Show, Generic)
+
+instance NoThunks (PrevGovActionIds era)
+instance Era era => NFData (PrevGovActionIds era)
+instance Default (PrevGovActionIds era)
+
+instance Era era => DecCBOR (PrevGovActionIds era) where
+  decCBOR =
+    decode $
+      RecD PrevGovActionIds
+        <! From
+        <! From
+        <! From
+        <! From
+
+instance Era era => EncCBOR (PrevGovActionIds era) where
+  encCBOR PrevGovActionIds {..} =
+    encode $
+      Rec (PrevGovActionIds @era)
+        !> To pgaPParamUpdate
+        !> To pgaHardFork
+        !> To pgaCommittee
+        !> To pgaConstitution
+
+toPrevGovActionIdsPairs :: (KeyValue e a, Era era) => PrevGovActionIds era -> [a]
+toPrevGovActionIdsPairs pga@(PrevGovActionIds _ _ _ _) =
+  let PrevGovActionIds {..} = pga
+   in [ "pgaPParamUpdate" .= pgaPParamUpdate
+      , "pgaHardFork" .= pgaHardFork
+      , "pgaCommittee" .= pgaCommittee
+      , "pgaConstitution" .= pgaConstitution
+      ]
+
+instance Era era => ToJSON (PrevGovActionIds era) where
+  toJSON = object . toPrevGovActionIdsPairs
+  toEncoding = pairs . mconcat . toPrevGovActionIdsPairs
+
+proposalsAddProposal ::
+  GovActionState era ->
+  PrevGovActionIds era ->
+  PrevGovActionIdsChildren era ->
+  Proposals era ->
+  Maybe (PrevGovActionIdsChildren era, Proposals era)
+proposalsAddProposal gas pgai pgac ps@(Proposals omap) =
+  let
+    parentIsEnacted = addToEnactedPrevGovActionIdsChildren gas pgai pgac
+    parentIsAnotherProposal = proposalsAddChild gas ps
+   in
+    case (parentIsEnacted, parentIsAnotherProposal) of
+      (Just updatedChildren, Just _updatedProposals) ->
+        -- The action is one of those that does not need a lineage. e.g. TreasuryWithdrawals
+        Just (updatedChildren, Proposals $ omap OMap.||> gas)
+      (Nothing, Nothing) ->
+        -- Parent not found, cannot accept this proposal
+        Nothing
+      (Just updatedChildren, Nothing) ->
+        -- Parent is an enacted action
+        Just (updatedChildren, Proposals $ omap OMap.||> gas)
+      (Nothing, Just (Proposals updatedOMap)) ->
+        -- Parent is another accepted proposal
+        Just (pgac, Proposals $ updatedOMap OMap.||> gas)
```

Comment:
It is not only about readability. By having the case statement on a tuple you are forcing `parentIsAnotherProposal` whenever it is unused, thus doing unnecessary computation.

```suggestion
    case parentIsEnacted of
      Just updatedChildren ->
        -- Parent is an enacted action
        Just (updatedChildren, Proposals $ omap OMap.||> gas)
      Nothing ->
        case parentIsAnotherProposal of
          -- Parent not found, cannot accept this proposal
          Nothing -> Nothing
          -- Parent is another accepted proposal
          Just (Proposals updatedOMap) ->
            Just (pgac, Proposals $ updatedOMap OMap.||> gas)
```

Reply:
:100: :raised_hands: Thank you!